R-packages/evalcast/R/geo_type_selector.R

```diff
@@ -1,50 +1,93 @@
+              msg = paste("A maximum of one geo_type may be provided to",
+                          "geo_type_selector. (Set geo_type = NULL and",
+                          "auto_select = TRUE to attempt auto-identification)"))
+  if (is.null(geo_type)) {
+    geo_types <- c("county", "hrr", "msa", "dma", "state", "nation")
+  } else {
+    if (auto_select) {
+      if (geo_type == "county") {
+        geo_types <- c("county", "state", "nation")
+      } else if (geo_type == "state") {
+        geo_types <- c("state", "county", "nation")
+      } else if (geo_type == "nation") {
+        geo_types <- c("nation", "state", "county")
+      } else {
+        geo_types <- geo_type
+      }
     } else {
-      assert_that(pred_card_types$has_chars, # actually is state
-                  pred_card_types$type_len == 2L,
-                  msg = paste("input was geo_type = ",
-                              "state, but predictions_cards do not entirely",
-                              "conform to this selection."))
+      geo_types <- geo_type
     }
   }
-  if (geo_type == "county") {
-    if (pred_card_types$type_len == 2L && pred_card_types$has_chars) {
-      geo_type <- "state" # switch this automatically
+  found_geo_type <- validate_geo_type(geo_types,
+                                     predictions_cards$geo_value,
+                                     auto_select)
+  if (is.null(found_geo_type)) {
+    if (is.null(geo_type)) {
+      stop(msg = paste("geo_type could not be auto-identified."))
     } else {
-      assert_that(!pred_card_types$has_chars, # actually is state
-                  pred_card_types$type_len == 5L,
-                  msg = paste("input was geo_type = ",
-                              "county, but predictions_cards do not entirely",
-                              "conform to this selection."))
+      stop(paste(
+        msg = "Provided geo_type was", geo_type,
+        "but format of some geo_values in predictions_cards do not match.")
+      )
+    }
+  }
+  else{
+    return(found_geo_type)
+  }
+}
+
+#' @param a vector of potential geo_types to check
+#' @param geo_values vector of strings to validate
+#'
+#' @return If `geo_values` conform to a type in `geo_types`, the first
+#'   conforming type is returned. Else return NULL.
+validate_geo_type <- function(geo_types, geo_values) {
+  assert_that(all(geo_types %in% c("county", "hrr", "msa", "dma", "state",
+                                    "nation")),
+              msg = "Unidentified geo_type")
+  geo_type_lengths <- unique(nchar(geo_values))
+  has_letters <- str_detect(geo_values, "[aA-zZ]")
+
+  if (length(geo_type_lengths) > 1) {
+    assert_that(any(geo_types %in% c("hrr", "msa", "dma")),
+                msg = paste("Provided geo_values vary in length, which is not",
+                            "allowed for provided geo_types:", geo_types))
+  }
+  return_type <- NULL
+  for (geo_type in geo_types) {
+    if (geo_type == "county") {
```

Comment:
Since these nested ifs have no corresponding else statements, consider combining them each into a single if.  However, I also understand why it might be more clear as written (branch on the `geo_type` then verify it works).  Your call.

Reply:
File removed.